### ISSUE TITLE
[discussion] potential bug in the scheduling path when one PreFilter plugin returns `Unschedulable`

### ISSUE THREAD

**Issue by Huang-Wei** [kubernetes/kubernetes#119770]
Move @sanposhiho 's https://github.com/kubernetes/kubernetes/pull/119769#pullrequestreview-1563700339 here as a new discussion thread:

Kensei pointed out a potential bug in the scheduling path when we hit Unschedulable result in one of PreFilter plugins:

> - pluginA returns Unschedulable in its PreFilter.
> - (Because pluginA returns unschedulable, RunPreFilterPlugins() is aborted at that point and pluginB PreFIlter isn't executed)
> - a Pod is claimed as unschedulable (failed by pluginA)
> - calling a pluginB's AddPod/RemovePod during preemption, sometimes will abort the whole preemption as the logic tries to read info from cycleState.

/sig scheduling
/kind bug

**Comment by sanposhiho** [comment]
One idea is to continue to run all PreFilter even if encountering Unschedulable from a plugin [here](https://github.com/kubernetes/kubernetes/blob/2c6c4566eff972d6c1320b5f8ad795f88c822d09/pkg/scheduler/framework/runtime/framework.go#L657).
It isn't a waste of time because the preemption will happen after that and all PreFilter data is needed for avoiding this bug + performant preemption.

**Comment by Huang-Wei** [comment]
@sanposhiho by looking into PreFilter() implementation of all in-tree plugins, it seems none of them returns `Unschedulable` . So we're lucky as this bug won't impact vanilla kube-scheduler that much (one edge case is PreFilter plugins may return `Error`, but that would only impact one round of scheduling). You're welcome to take a second look.

```
pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go
pkg/scheduler/framework/plugins/volumezone/volume_zone.go
pkg/scheduler/framework/plugins/nodevolumelimits/csi.go
pkg/scheduler/framework/plugins/nodevolumelimits/non_csi.go
pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go
pkg/scheduler/framework/plugins/noderesources/fit.go
pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
pkg/scheduler/framework/plugins/volumebinding/volume_binding.go
pkg/scheduler/framework/plugins/podtopologyspread/filtering.go
pkg/scheduler/framework/plugins/nodeports/node_ports.go
pkg/scheduler/framework/plugins/interpodaffinity/filtering.go
```

Rethink it a bit, given that returning `Unschedulable` in PreFilter() would abort the whole PreFilter call chain, and leave un-instantiated cycleState that may get error when calling PreFilterExtensions' Add/RemovePod - in short buggy workflow. I think PreFilter plugins should all return `UnschedulableAndUnresovable` to replace `Unschedulable`. Thoughts?

**Comment by Huang-Wei** [comment]
> One idea is to continue to run all PreFilter even if encountering Unschedulable from a plugin [here](https://github.com/kubernetes/kubernetes/blob/2c6c4566eff972d6c1320b5f8ad795f88c822d09/pkg/scheduler/framework/runtime/framework.go#L657).

I'm inclined to have all PreFilter() return `UnschedulableAndUnresovable` instead, and hence Preemption won't run.

**Comment by sanposhiho** [comment]
> it seems none of them returns Unschedulable

I didn't go through all plugins, but at least, DRA and volume_restriction return it..

- https://github.com/kubernetes/kubernetes/blob/2c6c4566eff972d6c1320b5f8ad795f88c822d09/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go#L559-L574
- https://github.com/kubernetes/kubernetes/blob/2c6c4566eff972d6c1320b5f8ad795f88c822d09/pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go#L182

**Comment by Huang-Wei** [comment]
Nope, I just meant `Unschedulable`, not `UnschedulableAndUnresovable`. `UnschedulableAndUnresovable` would block preemptin, so we'd never hit the bug.

**Comment by sanposhiho** [comment]
Ah, right. So, the default scheduler should be OK.

> I'm inclined to have all PreFilter() return UnschedulableAndUnresovable instead, and hence Preemption won't run.

So, it's already achieved in all in-tree plugins, right?
Then, what we _may_ need to mind about is out-of-tree plugins. If we assume there is a usecase that PreFilter() returns Unschedulable, then we still need to think a solution.

**Comment by sanposhiho** [comment]
Ah, no. We also need to care about this case even in the default scheduler:
https://github.com/kubernetes/kubernetes/blob/2c6c4566eff972d6c1320b5f8ad795f88c822d09/pkg/scheduler/framework/runtime/framework.go#L671

**Comment by sanposhiho** [comment]
So, I still basically prefer this:
https://github.com/kubernetes/kubernetes/issues/119770#issuecomment-1666390395

- continue to run PreFilter if a plugin return Unschedulable
  - it shouldn't happen in in-tree plugins though, custom plugins may return Unschedulable.
- continue to run PreFilter even if all Nodes are filtered out by `PreFilterResult` [here](https://github.com/kubernetes/kubernetes/blob/2c6c4566eff972d6c1320b5f8ad795f88c822d09/pkg/scheduler/framework/runtime/framework.go#L671)
  - it can happen in the default scheduler as well.

**Comment by sanposhiho** [comment]
Wait, one more thing.

@Huang-Wei 
>  it seems none of them returns Unschedulable . So we're lucky as this bug won't impact vanilla kube-scheduler that much

I may have missed something (please point out if I have), but I cannot find the code path where skipping preemption when UnschedulableAndUnresolvable is returned. So guessing currently we run the preemption even when UnschedulableAndUnresolvable is returned from PreFilter()?

**Comment by Huang-Wei** [comment]
@sanposhiho me as well... I think I mis-confused "PreFilter returns UnschedulableAndUnresovable" vs. "Filter returns UnschedulableAndUnresovable". Let me check the code more. But at least latest code doesn't distinguish them:

```go
	// Run "prefilter" plugins.
	preRes, s := fwk.RunPreFilterPlugins(ctx, state, pod)
	if !s.IsSuccess() {
		if !s.IsUnschedulable() {
			return nil, diagnosis, s.AsError()
		}
```

**Comment by sanposhiho** [comment]
Yes, I misunderstood as well.. According to the definition of `UnschedulableAndUnresolvable`, it makes sense not to run the preemption when UnschedulableAndUnresolvable is returned from PreFilter().
> 	// UnschedulableAndUnresolvable is used when a plugin finds a pod unschedulable and
	// other postFilter plugins like preemption would not change anything.

---

Do you have any other ideas apart from https://github.com/kubernetes/kubernetes/issues/119770#issuecomment-1666398214? (If not, I'll proceed with it.)
I'll prepare two PRs; one is to reproduce this problem in some tests, and the other is to fix the bug (for cherry-pick).

/assign

**Comment by Huang-Wei** [comment]
> Do you have any other ideas apart from https://github.com/kubernetes/kubernetes/issues/119770#issuecomment-1666398214? (If not, I'll proceed with it.)

Let's hold for a while. I do want to spend some time to see if UnschedulableAndUnresolvable works the same as Unschedulable in PreFilter. If it works the same, then the bug would impact both in-tree and out-of-tree plugins; if not, then in-tree plugins are not impacted.

**Comment by Huang-Wei** [comment]
@sanposhiho seems we have another regression... in old codebase, PreFilter's status `UnschedulableAndUnresolvable` was honored - different from `Unschedulable`, see the code in v1.25.12:

https://github.com/kubernetes/kubernetes/blob/v1.25.12/pkg/scheduler/schedule_one.go#L377-L387

However, the code was gone in v1.26. Let me find the curlsprit.

**Comment by sanposhiho** [comment]
Seems https://github.com/kubernetes/kubernetes/pull/110894 is the PR that removing ↓.

```go
		for _, n := range allNodes {
			diagnosis.NodeToStatusMap[n.Node().Name] = s
		}
```

When `UnschedulableAndUnresolvable` is returned from PreFilter, this part registered UnschedulableAndUnresolvable to all status and blocked preemption to run. We should get this back at least, or revert the PR.

**Comment by Huang-Wei** [comment]
@sanposhiho I think that's the regression PR. As the old/correct logic, although a bit verbose, is to apply the status returned from PreFilter to all nodes, and then UnschedulableAndUnresolvable is expected to be honored during default_preemption:

https://github.com/kubernetes/kubernetes/blob/2c6c4566eff972d6c1320b5f8ad795f88c822d09/pkg/scheduler/framework/preemption/preemption.go#L418-L423

**Comment by Huang-Wei** [comment]
@sanposhiho let's use a new issue to track the regression we spot [above](https://github.com/kubernetes/kubernetes/issues/119770#issuecomment-1666424312). (and once the issue is opened, remove the `kind/regression` label here as this issue exists in all versions)

Back to this issue, I'm fine with https://github.com/kubernetes/kubernetes/issues/119770#issuecomment-1666390395. That would be an no-op for vanilla scheduler (as none returns `Unschedulable` in PreFilter), while enabling out-of-tree plugins return `Unschedulable` w/o restrictions in PreFilter.

---

Some extra notes: the fix we proposed is only necessary when the issue is triggered, i.e., a scheduling workflow satisfying the following criteria:

- an out-of-tree pluginA returns `Unschedulable` in PreFilter
- an out-of-tree pluginB is placed after pluginA, and
  - it implements PreFilter, and write its own key to cycleState
  - it implements PreFilterExtensions#Add/RemovePod, and report error upon missing key

So I will bring this topic to next sig meeting to see how we proceed with it:

- provide a fix and cherrypick to all supported versions
- provide a fix but don't cherrypick
- or, maybe no fix? we may instruct plugin developers to avoid triggering the issue

**Comment by sanposhiho** [comment]
OK, let me have another issue.

**Comment by sanposhiho** [comment]
Created: https://github.com/kubernetes/kubernetes/issues/119782. (also the patch PR https://github.com/kubernetes/kubernetes/pull/119778)

---

We have two remaining issues. 
- when one PreFilter plugin returns Unschedulable
- when PreFilter plugin(s) filter out all nodes by PreFilterResult

Both could result in "not found data in cyclestate" error during the preemption.
The former case shouldn't happen in the default scheduler since we don't have PreFilter plugins returning Unschedulable.
But, the latter case may happen in the default scheduler.

Given both bugs are very similar, I'm wondering if we should have the fix for both in https://github.com/kubernetes/kubernetes/pull/119779, and cherry-pick it to all supported versions.
(PreFilterResult was introduced at 1.24.)

**Comment by alculquicondor** [comment]
Out of curiosity... what's a use case where you would return `Unschedulable` instead of `UnschedulableAndUnresolvable` in PreFilter.

I'm ok continuing with the rest of the plugins in the case of `Unschedulable`

**Comment by Huang-Wei** [comment]
I was curious on that as well... see https://github.com/kubernetes/kubernetes/issues/119770#issuecomment-1666394612

An example is the out-of-tree plugin [capacityscheduling](https://github.com/kubernetes-sigs/scheduler-plugins/tree/master/pkg/capacityscheduling) which reads the CR ElasticQuota in PreFilter (well, not strictly read from informer, it maintained a ElasticQuotaInfo like NodeInfo). If by adding the incoming Pod, the ElasticQuota's usage exceeds its max, it means it cannot be scheduled - and this logic is tangent with whether a Node can host or not, so PreFilter fits perfectly. Thus, the plugin chose to return `Unschedulable` in PreFilter, and then in PostFilter try preempting low-priority pods that resides in the same ElasticQuota.

**Comment by lance5890** [comment]
I have encounter this issue several times in k8s 1.25.8,  even in a single node, when we delete a sts pod(without deleting its pvc)，then the sts pod is pending, this will recover when we restart its kube-scheduler

**Comment by Huang-Wei** [comment]
> it seems like the kube-scheduler only recycle the Filter plugin, but without the RemovePod logic

It's not the same issue IMO. As the "RemovePod" referred to is a dryrun operation that's wired to preemption's execution path.

**Comment by lance5890** [comment]
> > it seems like the kube-scheduler only recycle the Filter plugin, but without the RemovePod logic
> 
> It's not the same issue IMO. As the "RemovePod" referred to is a dryrun operation that's wired to preemption's execution path.

maybe related https://github.com/kubernetes/kubernetes/issues/121866 and https://github.com/kubernetes/kubernetes/pull/116729

**Comment by Huang-Wei** [comment]
TL;DR for the latest status:

- **Fact:** preemption may not behave properly when some PreFilter returns `Unschedulable`
- **Lucky thing:** default-scheduler is not impacted as all in-tree scheduler plugins returns `UnschedulableAndUnresolvable` in PreFilter
- **Unlucky thing:** out-of-tree plugins or future in-tree plugins have reasonable requirements to return `Unschedulable` in PreFilter (example [here](https://github.com/kubernetes/kubernetes/issues/119770#issuecomment-1673954731)) and hence stomp on this bug

TODO:

- Provide a fix (#119779 is the most up-to-date one, cc @sanposhiho for confirmation)
- Reach a consensus to evaluate whether we should cherry-pick the fix

**Comment by sanposhiho** [comment]
Oops, https://github.com/kubernetes/kubernetes/pull/119779 slipped my mind.

**Comment by alculquicondor** [comment]
+1 on the feature request (not a fix)

Thus, it should not be cherry-picked.

**Comment by sanposhiho** [comment]
https://github.com/kubernetes/kubernetes/pull/119779 is merged, I think we can close it now if I don't miss any other scenarios.
